pipeline/data/lang_script.py

```diff
@@ -0,0 +1,319 @@
+    # to the base unit are typically diacritical marks that modify a base consonant and
+    # combine with it visually.
+    #
+    # For instance in Hindi क (ka) by itself implies /ka/, but adding the diacritic ि
+    # it becomes कि = /ki/.
+    #
+    # Examples: Devanagari, Bengali, Tamil, Thai, Khmer, Ethiopic
+    ABUGIDA = "abugida"
+
+    # A script where each symbol represents an entire syllable.
+    # Examples: Cherokee, Vai, Yi, Japanese Kana
+    SYLLABARY = "syllabary"
+
+    # A script where each symbol represents a word or morpheme, not individual sounds.
+    # Examples: Chinese Han characters, Japanese Kanji
+    LOGOGRAPHIC = "logographic"
+
+    # A script where characters are built from shapes representing phonetic features.
+    # Example: Hangul (Korean)
+    FEATURAL = "featural"
+
+
+class ScriptInfo(TypedDict):
+    name: str
+    type: ScriptType
+    bicameral: bool
+
+
+def is_script_phonemic(script_type: ScriptType):
+    """
+    Phonemic scripts are ones where the graphemes used, and their ordering directly
+    relates to their the sounds they make (phonemes). This is in contrast to
+    logographic scripts which use ideographs. One edge case is Hangul, which is featural.
+    It is treated as logographic even though it's technically phonemic due to the way the
+    characters are constructed from phonemic features (see #1101).
+    """
+    return script_type in {
+        ScriptType.ALPHABETIC,
+        ScriptType.ABJAD,
+        ScriptType.ABUGIDA,
+        ScriptType.SYLLABARY,
+    }
+
+
+def get_script_info(locale_str: str) -> Optional["ScriptInfo"]:
+    """
+    Take a BCP-47 locale code, and get the script info for it. This information is needed
+    for knowing what type of augmentations can be performed on the data.
+    """
+    # Load pyicu inline, as it only compiles on Linux.
+    from icu import Locale  # type: ignore[reportAttributeAccessIssue]
+
+    try:
+        icu_locale = Locale(locale_str)
+        maximized = Locale.addLikelySubtags(icu_locale)
+        script_code = maximized.getScript()
+
+        return scripts.get(script_code)
+
+    except Exception as exception:
+        print("Error getting script from locale", exception)
+
+    return None
+
+
+# This list was built from pulling CLDR data at an attempt to understand casing support.
+# https://github.com/unicode-org/cldr/tree/main/common/casing
+#
+# There are many more scripts available, and if needed support can be added here.
+# https://www.unicode.org/standard/supported.html
+#
+# To add a new script, look up information, generally from Wikipedia on the script
+# type and whether it is bicameral (e.g. cased with an upper and lower case).
+scripts: dict[str, ScriptInfo] = {
+    "Adlm": {
+        # 𞤀𞤤𞤳𞤵𞤤𞤫 𞤁𞤢𞤲𞤣𞤢𞤴𞤯𞤫 𞤂𞤫𞤻𞤮𞤤 𞤃𞤵𞤤𞤵𞤺𞤮𞤤
+        "name": "Adlam",
+        "type": ScriptType.ALPHABETIC,
+        "bicameral": True,
+    },
+    "Arab": {
+        # العربية
+        "name": "Arabic",
```

Comment:
Using a data class here instead of the fixed dictionary keys would make it easier to use and less error-prone.

Reply:
It's using a [`TypedDict`](https://typing.python.org/en/latest/spec/typeddict.html) which has type checking, so should be equivalent to a dataclass but with slightly different syntax: 

```
class ScriptInfo(TypedDict):
    name: str
    type: ScriptType
    bicameral: bool
```